### PR TITLE
Support annotations for type referenced record fields

### LINKED ISSUE #36409: Support annotations for type referenced record fields
**Description:**
$subject

**Describe your problem(s)**
```ballerina
public annotation IntConstraints Int on type, record field;

type IntConstraints record {|
    int minValue?;
|};

@Int {minValue: 0}
type PositiveInt int;

type Person record {
    PositiveInt age;
};
```
The record field `age` should contain the annotation data for `PositiveInt`

**Related Issues (optional):**
#35270

### PR DESCRIPTION
**Pull request by HindujaB**
## Purpose
$subject
Fixes #36409 

## Approach
Added relevant compiler FE, code-gen and runtime implementations to support annotations for the below type definitions.

## Samples
```ballerina
import testorg/sample as mod;

public annotation IntConstraints Int on type, record field;

type IntConstraints record {|
    int minValue?;
|};

@Int {minValue: 0}
type PositiveInt int;

@Int {minValue: 4}
type PositiveIntRef PositiveInt;

@Int {minValue: 5}
type PositiveIntRO PositiveInt & readonly;

@Int {minValue: 6}
type PositiveIntRef2 mod:PositiveInt;

type Person record {
    PositiveInt age;
};
```

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/split/constants/JvmRefTypeConstantsGen.java [hunk 1]
```diff
@@ -82,14 +87,25 @@ private void visitTypeRefTypeInitMethod() {
 
     private String generateTypeRefInits(BTypeReferenceType type) {
         String varName =
-                JvmConstants.TYPEREF_TYPE_VAR_PREFIX + Utils.encodeNonFunctionIdentifier(type.tsymbol.name.value);
+                JvmConstants.TYPEREF_TYPE_VAR_PREFIX + Utils.encodeNonFunctionIdentifier(type.getQualifiedTypeName());
```
[the remark this answers is not in the document]
Comment (warunalakshitha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/split/constants/JvmRefTypeConstantsGen.java:90]: Lets add a test for imported module ref types as well.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/split/JvmCreateTypeGen.java [hunk 2]
```diff
@@ -576,6 +576,14 @@ private AnonTypeHashInfo createLabelsForAnonTypeHashSwitch(Set<BIRTypeDefinition
         return new AnonTypeHashInfo(hashes, labels, labelFieldMapping);
     }
 
+    public void generateRefTypeConstants(List<BIRTypeDefinition> typeDefs, SymbolTable symbolTable) {
```
Comment (HindujaB) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/split/JvmCreateTypeGen.java:579]: The type-reference type constants need to be created for all reference type definitions, as they are referred to when called from another module.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmCodeGenUtil.java [hunk 3]
```diff
@@ -771,4 +771,8 @@ public static void castToJavaString(MethodVisitor mv, int fieldNameRegIndex, int
 
     private JvmCodeGenUtil() {
     }
+
+    public static String getRefTypeConstantName(BTypeReferenceType type) {
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmCodeGenUtil.java:775]: should this be part of jvmRefTypeconstantGenclass

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmSignatures.java [hunk 4]
```diff
@@ -363,7 +363,7 @@ public class JvmSignatures {
     public static final String STACK_FRAMES = "L" + STACK + ";";
     public static final String RECORD_GET = "(L" + STRING_VALUE + ";L" + OBJECT + ";)L" + OBJECT + ";";
     public static final String RECORD_GET_KEYS = "()[L" + OBJECT + ";";
-    public static final String RECORD_INIT = "(L" + TYPE + ";)V";
+    public static final String TYPE_PARAMETER = "(L" + TYPE + ";)V";
```
Comment (warunalakshitha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmSignatures.java:366]: Shall we move this in all Type* var name section

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmCodeGenUtil.java (hunk 3 above)
Comment (HindujaB) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/codegen/JvmCodeGenUtil.java:775]: We use this in `JvmTypeGen` as well. So, I kept this method in the commonplace.

**Review by warunalakshitha — APPROVED** [review]
LGTM

**Pull request merged**